File: PHP/NUBAN.php
<?php

/**
* Class to validate NUBAN
* To use, instantiate class with values and call the validate method
* @param string $bankCode, $accountNumber [Account number and corresponsing bank code for the account]
* @author Arowosegbe Ifeoluwa
* @return string Valid | Invalid
*/
class NUBAN
{
	protected $bankCode;
	protected $accountNumber;

	function __construct($bankCode, $accountNumber)
	{
		$this->bankCode = $bankCode;
		$this->accountNumber = $accountNumber;
	}

	public function validate()
	{
		if (strlen($this->bankCode) == 3 && strlen($this->accountNumber) == 10) {
			
			$nubanAccSerialNum = substr($this->accountNumber, 0, -1);
			$checkDigit = substr($this->accountNumber, -1);
			$nubanFormat = $this->bankCode . $nubanAccSerialNum;

			$nubanArray = str_split($nubanFormat);
			$algoDictionary = [3, 7, 3, 3, 7, 3, 3, 7, 3, 3, 7, 3];

			$checkSum = 0;

			foreach ($nubanArray as $key => $value) {
				$checkSum += ($value * $algoDictionary[$key]);
			}

			$validatedCheckDigit = 10 - ($checkSum % 10);
			$validatedCheckDigit = $validatedCheckDigit == 10 ? 0 : $validatedCheckDigit;

			return $checkDigit == $validatedCheckDigit ? 'Valid' : 'Invalid';	
			
		}else{
			throw new Exception("Invalid Bank Code or Account Number Length");			
		}
	}
}

File: Kotlin/Nuban.kt
package com.fairmoney.domain.bank

/**
 * NUBAN standards: https://www.cbn.gov.ng/Out/2020/PSMD/REVISED%20STANDARDS%20ON%20NIGERIA%20UNIFORM%20BANK%20ACCOUNT%20NUMBER%20(NUBAN)%20FOR%20BANKS%20AND%20OTHER%20FINANCIAL%20INSTITUTIONS%20.pdf
 */
class NubanValidator {

    private companion object {
        /**
         *  DMB: Deposit Money Banks
         */
        private const val DMB_BANK_CODE_LENGTH = 3

        /**
         * OFI: Other Financial Institutions
         */
        private const val OFI_BANK_CODE_LENGTH = 5
        private val ALGORITHM = intArrayOf(3, 7, 3, 3, 7, 3, 3, 7, 3, 3, 7, 3, 3, 7, 3)
    }

    fun validate(bankCode: String, accountNumber: String): Boolean {
        require(accountNumber.length == 10) { "Account number must be 10 digits long" }

        val formattedBankCode = when (bankCode.length) {
            3 -> "000$bankCode"
            5 -> "9$bankCode"
            else -> throw IllegalArgumentException("Bank code must be 3 or 5 digits")
        }

        val nubanSerialNumber = accountNumber.take(9)
        val checkDigit = Character.getNumericValue(accountNumber.last())

        val validatedCheckDigit = 10
                .minus(
                        (formattedBankCode + nubanSerialNumber)
                                .mapIndexed { index, char -> Character.getNumericValue(char) * ALGORITHM[index] }
                                .sum()
                                .rem(10)
                )
                .rem(10)
        
        return checkDigit == validatedCheckDigit
    }
}
